lib/node/runner/checker.go

```diff
@@ -155,13 +157,117 @@ func BallotNotFromKnownValidators(c common.Checker, args ...interface{}) (err er
 	return
 }
 
+// BallotCheckSYNC performs sync by considering sync condition.
+// And to participate in the consensus,
+// update the latestblock by referring to the database.
+func BallotCheckSYNC(c common.Checker, args ...interface{}) (err error) {
+	checker := c.(*BallotChecker)
+
+	is := checker.NodeRunner.Consensus()
+	b := checker.Ballot
+	if is.LatestBlock().Height >= b.Round().BlockHeight { // in consensus, not sync
+		return
+	}
+
+	if !isBallotAcceptYes(b) {
+		return
+	}
+
+	if !hasBallotValidProposer(is, b) {
+		return
+	}
+
+	if is.LatestBallot.H.Hash == "" {
+		is.LatestBallot = b
+	}
+
+	is.SaveNodeHeight(b.Source(), b.Round().BlockHeight)
+
+	var syncHeight uint64
+	var nodeAddrs []string
+	syncHeight, nodeAddrs, err = checker.NodeRunner.Consensus().GetSyncInfo()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if b.Round().BlockHeight == syncHeight {
+			checker.NodeRunner.Log().Debug("is.SetLatestBlock(blk)", "syncHeight", syncHeight)
+			is.LatestBallot = b
+		}
+	}()
+
+	if err = updateLatestBlockFromDatabase(is, checker.NodeRunner.storage, checker.NodeRunner.Log()); err != nil {
+		return
+	}
+
+	latestHeight := is.LatestBlock().Height
+	if latestHeight < syncHeight-1 { // request sync until syncHeight
+		checker.NodeRunner.Log().Debug("latestHeight < syncHeight-1", "latestHeight", latestHeight, "syncHeight", syncHeight)
+		is.StartSync(syncHeight, nodeAddrs)
+	} else {
+		if latestHeight == syncHeight-1 { // finish previous and current height ballot
+			var previousBlock *block.Block
+			previousBlock, err = finishBallot(
+				checker.NodeRunner.Storage(),
+				is.LatestBallot,
+				checker.NodeRunner.Consensus().TransactionPool,
+				checker.Log,
+				checker.NodeRunner.Log(),
+			)
+			is.SetLatestBlock(*previousBlock)
+			if err != nil {
+				return
+			}
+		}
+
+		var block *block.Block
+		block, err = finishBallot(
+			checker.NodeRunner.Storage(),
+			checker.Ballot,
+			checker.NodeRunner.Consensus().TransactionPool,
+			checker.Log,
+			checker.NodeRunner.Log(),
+		)
+		if err != nil {
+			return
+		}
+
+		checker.NodeRunner.Consensus().SetLatestBlock(*block)
+		checker.LocalNode.SetConsensus()
+		checker.NodeRunner.TransitISAACState(b.Round(), ballot.StateALLCONFIRM)
+		err = NewCheckerStopCloseConsensus(checker, "ballot got consensus")
+		return
+	}
+
+	return
+}
+
+func isBallotAcceptYes(b ballot.Ballot) bool {
+	return b.State() == ballot.StateACCEPT && b.Vote() == ballot.VotingYES
+}
+
+func hasBallotValidProposer(is *consensus.ISAAC, b ballot.Ballot) bool {
+	return b.Proposer() == is.SelectProposer(b.Round().BlockHeight, b.Round().Number)
+}
+
+func updateLatestBlockFromDatabase(is *consensus.ISAAC, st *storage.LevelDBBackend, log logging.Logger) error {
+	blk := block.GetLatestBlock(st)
+	if is.LatestBlock().Height < blk.Height {
+		is.SetLatestBlock(blk)
```

Comment:
Maybe we shouldn't have `SetLatestBlock` then... Or at the very least, it should accept the height, not a full block

Reply:
I want to know the reason.

Reply:
Sure! Sorry I should have made it clearer.
It's very error prone to have two ways of doing the same things. We have `block.GetLatestBlock` which accesses the storage, and `ISAAC.LatestBlock` which use a cached copy.

The issue here is that one might accidentally store a block and miss `SetLatestBlock`, or `SetLatestBlock` a wrong block, or `SetLatestBlock` the right block but not store it properly.
This comes from the fact that there is no link between `SetLatestBlock` and the storage. One way to do it would be to have `ISAAC.GetLatestBlock` just return `block.GetLatestBlock`, but I'm sure there is a reason we have this method in ISAAC. So we could instead change `SetLatestBlock` to accept the block height, and this way, we are sure that we can never the the wrong block / an inexistant block as the latest. It does not prevent the situation where we forget to call `SetLatestBlock`, but it reduces the risk of error.

Does that makes sense ?